src/concrete/ml/sklearn/base.py

```diff
@@ -1690,3 +1695,385 @@ def predict_proba(self, X: Data, fhe: Union[FheMode, str] = FheMode.DISABLE) ->
         y_logits = self.decision_function(X, fhe=fhe)
         y_proba = self.post_processing(y_logits)
         return y_proba
+
+
+# pylint: disable-next=invalid-name,too-many-instance-attributes
+class SklearnKNeighborsMixin(BaseEstimator, sklearn.base.BaseEstimator, ABC):
+    """A Mixin class for sklearn KNeighbors models with FHE.
+
+    This class inherits from sklearn.base.BaseEstimator in order to have access to scikit-learn's
+    `get_params` and `set_params` methods.
+    """
+
+    def __init_subclass__(cls):
+        for klass in cls.__mro__:
+            # pylint: disable-next=protected-access
+            if getattr(klass, "_is_a_public_cml_model", False):
+                _NEIGHBORS_MODELS.add(cls)
+                _ALL_SKLEARN_MODELS.add(cls)
+
+    def __init__(self, n_bits: Union[int, Dict[str, int]] = 3):
+        """Initialize the FHE knn model.
+
+        Args:
+            n_bits (int, Dict[str, int]): Number of bits to quantize the model. If an int is passed
+                for n_bits, the value will be used for quantizing inputs and weights. If a dict is
+                passed, then it should contain "op_inputs" and "op_weights" as keys with
+                corresponding number of quantization bits so that:
+                    - op_inputs : number of bits to quantize the input values
+                    - op_weights: number of bits to quantize the learned parameters
+                Default to 3.
+        """
+        self.n_bits: Union[int, Dict[str, int]] = n_bits
+
+        #: The quantizer to use for quantizing the model's weights
+        self._weight_quantizer: Optional[UniformQuantizer] = None
+        # In distance metric algorithms, there is no `weights` attribute because there isn't
+        # a training phase.
+        # Instead, we have `_X_fit` attribute, commonly used in scikit-learn, which stores the
+        # training set, in order to compute the pairewise Euclidean distance.
+        self._q_X_fit_quantizer: Optional[UniformQuantizer] = None
+        self._q_X_fit: numpy.ndarray
+
+        BaseEstimator.__init__(self)
+
+    def _set_onnx_model(self, test_input: numpy.ndarray) -> None:
+        """Retrieve the model's ONNX graph using Hummingbird conversion.
+
+        Args:
+            test_input (numpy.ndarray): An input data used to trace the model execution.
+        """
+        # Check that the underlying sklearn model has been set and fit
+        assert self.sklearn_model is not None, self._sklearn_model_is_not_fitted_error_message()
+
+        self.onnx_model_ = hb_convert(
```

Comment:
hmm do we use this? I think we coded everything from scratch so I don't get why we extract the onnx here.

Reply:
I'll close the comment according to this discussion: https://zama-ai.slack.com/archives/C0455QDGT6C/p1695131751281249